### PR TITLE
Fix ByteBuf reference counting

### PR DESCRIPTION
**Pull request by adleong**
The Netty ByteBufs used in Linkerd's H2 implementation are manually reference counted.  There are a number of situations where we fail to call `release()` the correct number of times which results in a ByteBuf leak.  

We make the following changes to prevent these leaks:

* Whenever we would discard the contains of a frame queue, we instead drain the queue and release each of those frames.
* No longer call `release()` on a Netty frame after writing it, instead relying on the H2 frame's releaser to be responsible for releasing the underlying Buf.
* No longer call `retain()` on the underlying Buf when constructing an H2 frame from a Netty frame (except in the case of array backed Bufs).
* Ensure we release frames before dropping them when errors occur.
* When decoding a gRPC message, copy the data out of the frame before releasing it.
* Remove a spurious call to `release()` when adding a trailers frame to a stream.

The end result is that there are only a few cases where `retain()` and `release()` are called:

* When passing ByteBuf to Netty to be written, we `retain()` it first so that it isn't freed when Netty writes it and calls `release()`.  This is desirable because we may be using the ByteBuf elsewhere (for example, in the case of retries) and want to be in control of when the ByteBuf is freed.
* After sending a frame to Netty to be written, we call `release()` on the H2 frame.  If this frame is being used in multiple places then it will be a RefCountedFrame and this call to `release()` will only trigger the frame's releaser when the reference count reaches 0.
* When a RefCountedFrame's reference count reaches 0 or when a non-RefCounted Frame is released, it's releaser will call `release()` on the underlying ByteBuf (and update flow control).  If all Netty writes are complete, this is expected to reduce the ByteBuf's reference count to 0 and free it.
* If a frame would be dropped instead of being sent to Netty for writing, it must be manually `released()`.

### CODE REVIEW

File: finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/Stream.scala [hunk 1]
```diff
@@ -68,6 +68,17 @@ object Stream {
   // performed).  This would require some form of buffering, ideally
   // in the dispatcher and server stream
 
+  /**
+   * Read all frames from the queue and release them.  This can be called on a queue after
+   * calling q.fail(e, discard = false) as a safe alternative to q.fail(e, discard = true).
+   */
+  def drainFrameQueue(q: AsyncQueue[Frame]): Future[Nothing] = {
+    q.poll().flatMap { f =>
```
Comment (olix0r) [inline finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/Stream.scala]: consider including `q.fail(e, discard = false)` in here -- since this is never called without also failing immediately prior afaict.  then can be renamed to `failFrameQueue`

Comment (adleong) [reply finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/Stream.scala]: good idea

**Comment by adleong** [comment]
By default Netty samples 1% of buffers to check for leaks and logs if leaks are detected.  There are a number of scenarios (particularly when using gRPC) that trigger this leak detection on master that no longer trigger it after this change.

http://netty.io/wiki/reference-counted-objects.html#wiki-h3-11

File: router/h2/src/main/scala/io/buoyant/router/h2/ClassifierFilter.scala [hunk 2]
```diff
@@ -135,9 +135,7 @@ class ClassifierFilter(classifier: H2Classifier) extends SimpleFilter[Request, R
                 // replace it with a new data frame without that flag.
                 // otherwise, the trailers we add to the stream will never
                 // be reached.
-                val frame2 = Frame.Data(frame.buf, eos = false)
-                // release the old data frame, since it will be replaced.
-                frame.release()
```
Comment (hawkw) [inline router/h2/src/main/scala/io/buoyant/router/h2/ClassifierFilter.scala]: As we discussed previously, releasing the frame here was almost certainly incorrect. This LGTM 👍

File: router/h2/src/e2e/scala/io/buoyant/router/h2/FlowControlEndToEndTest.scala [hunk 3]
```diff
@@ -2,11 +2,9 @@ package io.buoyant.router
 package h2
 
 import com.twitter.finagle.buoyant.h2._
-import com.twitter.io.Buf
 import com.twitter.logging.Level
-import com.twitter.util.{Future, Promise}
+import com.twitter.util.Promise
 import io.buoyant.test.FunSuite
-import scala.collection.mutable.ListBuffer
```
Comment (hawkw) [inline router/h2/src/e2e/scala/io/buoyant/router/h2/FlowControlEndToEndTest.scala]: Were these imports unused? This looks like an unrelated change at a glance?

File: grpc/runtime/src/main/scala/io/buoyant/grpc/runtime/Stream.scala [hunk 4]
```diff
@@ -22,8 +22,22 @@ object Stream {
     def close(): Future[Unit]
   }
 
+  private def failAndDrainQueue[T](q: AsyncQueue[Releasable[T]], e: Throwable): Unit = {
```
Comment (hawkw) [inline grpc/runtime/src/main/scala/io/buoyant/grpc/runtime/Stream.scala]: Would be nice if there was a short comment here explaining why this is different from calling `q.fail(..., discard = true)` and why it's used instead.

File: finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/Stream.scala [hunk 5]
```diff
@@ -68,6 +68,22 @@ object Stream {
   // performed).  This would require some form of buffering, ideally
   // in the dispatcher and server stream
 
+  /**
+   * Fail the queue and release all of its Frames.  This can be called as a safe alternative to
+   * q.fail(e, discard = true).
+   */
+  def failAndDrainFrameQueue(q: AsyncQueue[Frame], e: Throwable): Unit = {
```
Comment (hawkw) [inline finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/Stream.scala]: Seems like there's some duplication between this function and the `failAndDrainQueue` function in `io.buoyant.grpc.runtime.Stream`; is there any reasonable way to factor it out so we can use the same code in both places?

File: router/h2/src/e2e/scala/io/buoyant/router/h2/RouterEndToEndTest.scala [hunk 6]
```diff
@@ -205,5 +205,4 @@ class RouterEndToEndTest
       await(router.close())
     }
   }
-
```
Comment (hawkw) [inline router/h2/src/e2e/scala/io/buoyant/router/h2/RouterEndToEndTest.scala]: Unrelated change?

File: router/h2/src/e2e/scala/io/buoyant/router/h2/FlowControlEndToEndTest.scala (hunk 3 above)
Comment (adleong) [reply router/h2/src/e2e/scala/io/buoyant/router/h2/FlowControlEndToEndTest.scala]: yup, an unrelated "clean-as-I-go" change

File: router/h2/src/e2e/scala/io/buoyant/router/h2/RouterEndToEndTest.scala (hunk 6 above)
Comment (adleong) [reply router/h2/src/e2e/scala/io/buoyant/router/h2/RouterEndToEndTest.scala]: yup, an unrelated "clean-as-I-go" change

File: finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/Stream.scala (hunk 5 above)
Comment (adleong) [reply finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/Stream.scala]: Yeah, the functions are basically identical except that one operates on Frames and the other operates on Releasables.  We could consolidate them by having Frame and Releasable implement a common trait or by having `failAndDrainQueue` take a releaser function as an input.  But, given the small size of this method, I think I'd rather avoid adding an abstraction.

Comment (hawkw) [reply finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/Stream.scala]: Oh, I didn't realise the functions were operating on different types; carry on, then!
